### ISSUE TITLE
[Feature Request] panic when user try get recursive lock.

### ISSUE THREAD

**Issue by cathaysia** [Amanieu/parking_lot#438]
I had notice that rwlock is not a recursive lock:

> This lock uses a task-fair locking policy which avoids both reader and writer starvation. This means that readers trying to acquire the lock will block even if the lock is unlocked when there are writers waiting to acquire the lock. Because of this, attempts to recursively acquire a read lock within a single thread may result in a deadlock.

That is to say, I cannot acquire the read lock and write lock at the same time in one thread.

On Linux, trying to acquire a recursive lock on a non-recursive lock causes the program to crash. Should packing_lot also have this behavior? Crashing the program early can reduce debugging time (especially for problems like deadlocks)

**Comment by timnewsham** [comment]
I came here to make the same feature request:
**There should be an optional feature to panic on duplicate read locks from the same thread.**
This would be useful when running test cases. As it stands now, if you wrote a test case where a thread grabbed the same read lock twice it will pass just fine, unless you manage to get a concurrent write lock to occur. This is not practical in testing.

**Comment by Amanieu** [comment]
I'm not opposed to having this as an optional feature, but you need to be aware that this will severely hurt performance. Essentially this will require a thread-local `HashSet` of all read locks that the current thread already owns. And it won't catch lock cycles across multiple threads.

**Comment by cathaysia** [comment]
Yes, it will hurt performance. But it's very useful for debug. ;)

**Comment by Amanieu** [comment]
I'm happy to accept a PR if you are willing to write one.

**Comment by cathaysia** [comment]
I am glad to hear this. I will try to take time to do this.

**Comment by cathaysia** [comment]
Closing this issue because there has been no non-automation comment for 663 days.

**Comment by cathaysia** [comment]
🤖 Closing this issue because there has been no non-automation comment for 712 days.

closed by [#54](https://github.com/cathaysia/cathaysia/actions/runs/32094373472)
